### PR TITLE
[kvstore] passed tests on dist_sync for various cnns

### PR DESCRIPTION
**Pull request by mli**

### CODE REVIEW

**Comment by tqchen** [comment]
Consider still use ndarray on server. But get the pointer representation by
data()

**Comment by mli** [comment]
it's hard to make sure all these ndarrays are deleted before engine is
shutting down base on the current ps-lite implementation

**Comment by tqchen** [comment]
If ps lite is singleton, use get Ref to get engine and storage ref into ps
lite server

**Comment by mli** [comment]
but why uses ndarray? to avoid frequently var create/delete for better performance?

**Comment by tqchen** [comment]
Yes. For that thread in general.

**Comment by tqchen** [comment]
Also eventually mshadow should give aligned allocation and vectorization
speed ups which propagates to nd

File: python/mxnet/model.py [hunk 1]
```diff
@@ -221,15 +222,11 @@ def _train_multi_device(symbol, ctx, input_shape,
...
-        'local' : multi-devices on a single machine, will automatically
-           choose one from 'local_update_cpu', 'local_allreduce_cpu', and
-          'local_allreduce_device'
-       'dist_sync' : multi-machines with BSP
-       'dist_async' : multi-machines with partical asynchronous
+    kvstore : KVStore
+        The KVStore
 
-       In default uses 'local', often no need to change for single machiine.
+    update_on_kvstore : bool
```
Comment (tqchen) [inline python/mxnet/model.py]: This is not needed?

File: python/mxnet/model.py [hunk 2]
```diff
@@ -171,9 +171,10 @@ def _create_kvstore(kvstore, num_device, arg_params):
 def _train_multi_device(symbol, ctx, input_shape,
                         arg_params, aux_params,
                         begin_round, end_round, optimizer,
+                        kvstore, update_on_kvstore,
```
Comment (tqchen) [inline python/mxnet/model.py]: Added the update_on_kvstore back, which is not needed

**Comment by tqchen** [comment]
I made a few comments. I think this can be merged in once the comments are fixed. Would be nice if we can use NDArray to manage the space on server side.

File: python/mxnet/model.py (hunk 1 above)
Comment (mli) [inline python/mxnet/model.py]: it is needed. kvstore is now constructed earlier.

**Comment by mli** [comment]
i'll merge it now. these codes needed to be changed anyway since i'm changing the ps-lite interface

**Comment by tqchen** [comment]
I see. Another reason for using NDArray by default is it helps you to resolve write dep tracking. Which could be nice way to resolve some of the concurrent writing issues of server side dep scheduling(which I assume ps-lite took some part). Otherwise the update need to append WaitToWrite/Read, which might not be perfect way to resolve the deps

**Pull request merged**